[STC-5] GitHub repositories can be added/deleted to ScripTooCloud

jormsby2 opened the pull request:
Created service for cloning and deleting repositories.

Two views added to access this service. Manage repositories view for deleting/viewing repos. Clone repositories view for adding repositories

Clone repositories view accepts URL input, URL is processed by javascript and hidden fields are then filled for the model backed by hibernate validator (not sure if this is the best approach but it looks nice?)

Unit tests for service

jormsby2 committed:
[STC-5] Clone Service Finished

jormsby2 committed:
[STC-5] Finished unit tests

jormsby2 committed:
[STC-5] Formatting

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/config/MvcConfig.java:
```diff
@@ -1,9 +1,13 @@
+
 @Configuration
-@ComponentScan("edu.asu.diging.simpleusers.web")
+@ComponentScan({"edu.asu.diging.simpleusers.web","edu.asu.diging.scriptoocloud.core"})
```
`edu.asu.diging.scriptoocloud` is already scanned in the root context. It should not be necessary here.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/config/SecurityContext.java:
```diff
@@ -35,8 +35,8 @@ protected void configure(HttpSecurity http) throws Exception {
                 .antMatchers("/", "/resources/**",
                        "/register").permitAll()
                 // The rest of the our application is protected.
-                .antMatchers("/users/**", "/admin/**").hasRole("ADMIN")
```
Is there a reason you removed there?

jormsby2 replied:
I swapped the position of "USER" and "ADMIN" such that the new code was " .anyRequest().hasRole("ADMIN") ". I was having issues with the admin account not being able to access my views in /github/** until i did this.

jdamerow replied:
Now you just made everything only accessible for admins and nothing for users, `.antMatchers().hasRole("USER")` doesn't make much sense, it just doesn't give users access to anything. I would put any admin related actions under `/admin`, e.g. `/admin/github` and open that url only for admins.
`.antMatchers("/users/**", "/admin/**").hasRole("ADMIN")`.
basically says, any URL that starts with `/users` or `/admin` will be only accessible by users with the admin role. If you really want to keep `/github`, you can just add that to that line. However, it will get harder to keep track of which urls can be accessed by whom as your app grows. Prefixes help a lot with this.

`.anyRequest().hasRole("ADMIN")` says, everything that hasn't been mentioned above is only accessible by admins.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/model/Repository.java:
```diff
@@ -0,0 +1,35 @@
+
+import java.time.ZonedDateTime;
+
+public interface Repository {
```
I would not call this `Repository`. It'll make it easy to confuse with Spring Data repositories. How about `GitRepository`?

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/model/impl/RepositoryImpl.java:
```diff
@@ -0,0 +1,103 @@
+import edu.asu.diging.scriptoocloud.core.model.Repository;
+
+@Entity
+public class RepositoryImpl implements Repository {
```
same as interface

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/model/impl/RepositoryImpl.java:
```diff
@@ -0,0 +1,103 @@
+    private String owner;
+    
+    @NotEmpty(message = "Repository name required")
+    private String repo;
```
what does this hold? By repository name you mean what? the variable name should be selfexplanatory.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/model/impl/RepositoryImpl.java:
```diff
@@ -0,0 +1,103 @@
+    
+    private String url;
+    private ZonedDateTime creationDate;
+    private String Requester;
```
should be lowercase

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/CloneRepository.java:
```diff
@@ -0,0 +1,22 @@
+import edu.asu.diging.scriptoocloud.core.forms.CloneForm;
+import edu.asu.diging.scriptoocloud.core.model.Repository;
+
+public interface CloneRepository {
```
Is that a service or manager? If so, its name should indicate that.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/CloneRepositoryImpl.java:
```diff
@@ -0,0 +1,129 @@
+
+@Transactional
+@Service
+public class CloneRepositoryImpl implements CloneRepository{
```
same as above

jdamerow replied:
I would create a second class for all the code that deals with the file system and then have this class call. Single responsibility principle, only be responsible for one thing.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/CloneRepositoryImpl.java:
```diff
@@ -0,0 +1,129 @@
+public class CloneRepositoryImpl implements CloneRepository{
+
+    @Autowired 
+    private ClonedRepository clonedRepository;
```
here it gets really confusing with the names. What's the difference between `CloneRepository` and `ClonedRepository`?

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/CloneRepositoryImpl.java:
```diff
@@ -0,0 +1,129 @@
+    private String requester;
+    private String url;
+    private ZonedDateTime creationDate;
+    private String drive = "C:/";
```
Big, big, no no! ;) These are class variables that you set in your methods. There is only one object of this class (default unless you tell Spring otherwise), which means if you have two requests coming in at the same time, the second will overwrite the values of the first.

jdamerow replied:
Also, `drive` needs to be coming from the config file. I could not run this code, as I am on a Mac.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/CloneRepositoryImpl.java:
```diff
@@ -0,0 +1,129 @@
+        requester = ((IUser)SecurityContextHolder.getContext()
+                            .getAuthentication().getPrincipal()).getUsername();
+        creationDate = ZonedDateTime.now();                                   
+        url = host + "/" + owner + "/" +repo;
```
no need for these variables to be class variables

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/CloneRepositoryImpl.java:
```diff
@@ -0,0 +1,129 @@
+        host = cloneForm.getHost();
+        owner = cloneForm.getOwner();
+        repo = cloneForm.getRepo();
+        requester = ((IUser)SecurityContextHolder.getContext()
```
I would pass the user information as a parameter to decouple the business logic from spring security

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/CloneRepositoryImpl.java:
```diff
@@ -0,0 +1,129 @@
+                deleteDirectoryContents(parentDirectory);
+            }
+        }
+        catch(InvalidDataAccessResourceUsageException e){}
```
do not swallow exceptions (we'll talk about it today actually).

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/CloneRepositoryImpl.java:
```diff
@@ -0,0 +1,129 @@
+            .call().getRepository().close();
+        }
+        catch(GitAPIException e){
+            File parentDirectory = new File(drive + requester + "_" + owner + "_" + repo);
```
exceptions need to be logged

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/CloneRepositoryImpl.java:
```diff
@@ -0,0 +1,129 @@
+        catch(GitAPIException e){
+            File parentDirectory = new File(drive + requester + "_" + owner + "_" + repo);
+            deleteDirectoryContents(parentDirectory);
+            return "redirect:/clone" + "?badurl"; 
```
since this is a business logic class, it should not return view names.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/CloneRepositoryImpl.java:
```diff
@@ -0,0 +1,129 @@
+
+        clonedRepository.save(repository);
+        
+        return "redirect:/clone" + "?success"; 
```
same as above

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/CloneRepositoryImpl.java:
```diff
@@ -0,0 +1,129 @@
+            deleteDirectoryContents(parentDirectory);
+        }
+    
+        return "redirect:/listrepos" + "?success"; 
```
same as above

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/CloneRepositoryImpl.java:
```diff
@@ -0,0 +1,129 @@
+        {
+            file.delete();
+        }
+    }
```
`{` should be on the previous lines (see our code style)

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/CloneRepositoryImpl.java:
```diff
@@ -0,0 +1,129 @@
+        }
+    }
+
+}
```
methods need javadoc

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneRepositoryController.java:
```diff
@@ -0,0 +1,57 @@
+public class CloneRepositoryController {
+
+    @Autowired
+    CloneRepository cloneService;
```
should be private

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneRepositoryController.java:
```diff
@@ -0,0 +1,57 @@
+    @Autowired
+    CloneRepository cloneService;
+ 
+    @RequestMapping(value = "/listrepos", method = RequestMethod.GET)
```
how about `/repositories/list`?

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneRepositoryController.java:
```diff
@@ -0,0 +1,57 @@
+        return  "github/ListRepos";
+    }
+
+    @RequestMapping(value = "/listrepos/{host}/{owner}/{repo}/{requester}", method = RequestMethod.POST)
```
This method is for deleting repositories already added, right? you should use their id to delete them. Also, this method should be in its own controller.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneRepositoryController.java:
```diff
@@ -0,0 +1,57 @@
+                        @PathVariable("owner") String owner,
+                        @PathVariable("repo") String repo,
+                        @PathVariable("requester") String requester,
+                        @ModelAttribute("clone") CloneForm cloneForm, BindingResult result) 
```
is the `cloneForm` used?

jormsby2 replied:
The cloneform was for user inputs

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneRepositoryController.java:
```diff
@@ -0,0 +1,57 @@
+       return cloneService.deleteRepo(host + "/" + owner + "/" + repo, requester, owner, repo);        
+    }
+  
+    @RequestMapping(value = "/clone", method = RequestMethod.GET)
```
since only admins should be allowed to do this, this needs to be at a restricted url

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneRepositoryController.java:
```diff
@@ -0,0 +1,57 @@
+    public String clone(@Valid @ModelAttribute("clone") CloneForm cloneForm, BindingResult result, Model model) 
+                             throws GitAPIException {                                          
+        if(result.hasErrors()){
+            model.addAttribute("clone", new CloneForm());
```
If there are errors, you probably want to return the existing `cloneForm` so the user doesn't have to fill everything out again.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneRepositoryController.java:
```diff
@@ -0,0 +1,57 @@
+    }
+    
+    
+}
```
there should be one controller for each action. One for adding repositories, one for deleting them, one for listing them all.

jdamerow commented on scriptoocloud/src/main/webapp/WEB-INF/views/github/ListRepos.html:
```diff
@@ -0,0 +1,46 @@
+
+<h2>Repositories</h2>
+<ul class="list-group">
+ <li th:each="Repository : ${repo}"  class="list-group-item clearfix">
```
since we lowercase variables in java, I would lowercase them here too.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/CloneRepositoryImplTest.java:
```diff
@@ -0,0 +1,103 @@
+    }   
+  
+    
+
```
there should be several use cases for every method (in most cases), e.g. cloneRepo finishes successfully and failed;

jormsby2 committed:
[STC-5] Pull request fixes, unit testing not finished

jormsby2 committed:
[STC-5] Unit tests

jormsby2 committed:
[STC-5] jgitservice now deletes failed clone directory

jormsby2 committed:
[STC-5] Controller and form bug fixes

jormsby2 committed:
[STC-5] Refactored services

jormsby2 committed:
[STC-5] Unit tests

jormsby2 committed:
[STC-5] Formatting

jormsby2 commented:
I think I resolved the issues. I'm a bit uncertain how to create more tests then what I have here though.

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/data/GitRepositoryRepository.java:
```diff
@@ -0,0 +1,18 @@
+    
+    public GitRepositoryImpl findBygitRepositoryOwner(String repo);
+    
+    public GitRepositoryImpl findBygitRepositoryName(String owner);
```
"git" should be capitalized in the above three methods

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/DeleteGitRepositoryServiceImpl.java:
```diff
@@ -0,0 +1,51 @@
+        File file = new File(path + gitRepository.getRequester() + "_" + gitRepository.getGitRepositoryOwner() 
+                                + "_" + gitRepository.getGitRepositoryName());
+        deleteDirectoryContents(file);
+    }
```
What is the difference between this delete method and the one in the repository service?

jdamerow replied:
the repository service should be responsible for deleting (this class shouldn't have a dependency to GitRepositoryRepository). This class would probably be your file system connector that makes sure files are deleted for a given id/repository object.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/DeleteGitRepositoryServiceImpl.java:
```diff
@@ -0,0 +1,51 @@
+    
+    @Override
+    public void deleteDirectoryContents(File file){
+        File directorycontents[];   
```
declare where needed (in line 40)

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryService.java:
```diff
@@ -0,0 +1,83 @@
+        String repositoryName = cloneForm.getRepo();
+        String requester = user.getUsername();
+        ZonedDateTime creationDate = ZonedDateTime.now();                                   
+        String url = host + "/" + owner + "/" + repositoryName;
```
is host the server host (e.g. github.com)? If someone is not hosting their repository with GitHub, the url might have a different format (e.g. hostname/reponame). Since we for now focus on GitHub, the host should be a configuration parameter and not something the user has to enter. Or if you want it to work with other hosting services as well, just make it a url (no owner, reponame).

jdamerow replied:
also, why do you ask the user to enter a URL if you don't use it?

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryService.java:
```diff
@@ -0,0 +1,83 @@
+            gitRepositoryJpa.save(gitRepository);
+        }
+        catch(IllegalArgumentException e){
+            throw new IllegalArgumentException(e);
```
either keep throwing the `IllegalArgumentException` e or wrap it in a custom exception (which is probably what you want to do, the upper layers just need to know that the repository couldn't be cloned). Wrapping an exception in the same type doesn't make sense and doesn't add any value. When is that exception thrown? When what is passed?

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/JgitServiceImpl.java:
```diff
@@ -0,0 +1,32 @@
+class JgitServiceImpl implements JgitService{
+
+    @Autowired
+    DeleteGitRepositoryService deleteGitRepositoryService;
```
should be private

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/JgitServiceImpl.java:
```diff
@@ -0,0 +1,32 @@
+            .setDirectory(new File(filename))
+            .call().getRepository().close();
+        }
+        catch(GitAPIException e){
```
please adjust your eclipse code style, `catch` goes on the previous line

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/JgitServiceImpl.java:
```diff
@@ -0,0 +1,32 @@
+        }
+        catch(GitAPIException e){
+            deleteGitRepositoryService.deleteDirectoryContents(new File(filename));
+            throw new IllegalArgumentException(e);
```
why is that an `IllegalArgumentException`? this needs to be a proper checked exception. This is a case that the upper classes need to handle.

jormsby2 replied:
I'm not exactly sure what to do here? I threw an IllegalArgumentException to decouple the controller from jgit. What would be your recommendation here?

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneGitRepositoriesController.java:
```diff
@@ -0,0 +1,64 @@
+public class CloneGitRepositoriesController {
+    
+    @Autowired
+    GitRepositoryManager gitRepositoryManager;
```
private

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneGitRepositoriesController.java:
```diff
@@ -0,0 +1,64 @@
+    @Value("${git.repositories.path}")
+    private String path;
+  
+    private static Logger logger = LoggerFactory.getLogger(CloneGitRepositoriesController.class);    
```
should not be static; you have only one object of the controller anyways.

jdamerow replied:
and instead of passing `CloneGitRepositoriesController.class`, I would pass `getClass()` because
a) If you or someone else copies this line into another class, it does not need to be adjusted (less likely you end up with log messages for the wrong class)
b) for controllers, it's less likely but if you inherit from a class that declares a logger with `getClass()`, your logger will use the correct class for logging.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneGitRepositoriesController.java:
```diff
@@ -0,0 +1,64 @@
+        }                 
+        
+        IUser user = (IUser)SecurityContextHolder.getContext()
+                            .getAuthentication().getPrincipal();       
```
just add `Principal principal` and you should be good. Although, you might have to call the usermanager to get a user object. But it would decouple the controller from Spring Security.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneGitRepositoriesController.java:
```diff
@@ -0,0 +1,64 @@
+        }
+        catch(IllegalArgumentException e){
+            logger.error("No git repository found at provided URL" 
+                            + path + user + cloneForm.getOwner() + cloneForm.getRepo());
```
this would print
`No git repository found at provided URLpath/added/@UserImpl...myusermyrepo`
You can either add blanks or use `String.format(...)` to make it easier readable.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneGitRepositoriesController.java:
```diff
@@ -0,0 +1,64 @@
+            return "redirect:/repositories/clone" + "?badurl";
+       }
+       
+       return "redirect:/repositories/clone" + "?success";
```
Instead of this, I would inject `RedirectAttributes` into the method and then add flash attributes. Flash attributes will be passed to the next page, so in your view template, you can check for these attributes to show error/success messages. Also, this is not a proper url parameter, it needs a `=something` or `something=` (just use flash attributes ;)

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/DeleteGitRepositoryController.java:
```diff
@@ -0,0 +1,23 @@
+    @RequestMapping(value = "/repositories/delete/{id}", method = RequestMethod.POST)
+    public String deleteRepo(@PathVariable("id") Long id) {   
+        gitRepositoryManager.deleteRepository(id);
+        return "redirect:/repositories/list" + "?success";            
```
same as above

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/ListGitRepositoriesController.java:
```diff
@@ -0,0 +1,27 @@
+    @RequestMapping(value = "/repositories/list", method = RequestMethod.GET)
+    public String listRepos(Model model) {
+        ArrayList<GitRepository> repositories = gitRepositoryManager.listRepositories();
+        model.addAttribute("repos", repositories);
```
merge with line 22

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DeleteGitRepositoryServiceImplTest.java:
```diff
@@ -0,0 +1,42 @@
+    @InjectMocks
+    private DeleteGitRepositoryServiceImpl serviceToTest;
+    
+    private File file;
```
does not need to be a class variable

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DeleteGitRepositoryServiceImplTest.java:
```diff
@@ -0,0 +1,42 @@
+        Assertions.assertEquals(true, file.exists());
+        serviceToTest.deleteDirectoryContents(file);
+        Assertions.assertEquals(false, file.exists());
+    }   
```
needs a test case for when the file does not exist

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryManagerServiceTest.java:
```diff
@@ -0,0 +1,56 @@
+    private JgitService jGitService;
+    
+    @Mock
+    GitRepositoryRepository gitRepositoryJpa;
```
should be private (it's not that important here, but it's a good practice to keep)

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryManagerServiceTest.java:
```diff
@@ -0,0 +1,56 @@
+    
+    @Test
+    public void test_cloneRepository_clone() {
+        Mockito.when(gitRepositoryJpa.findByUrl("")).thenReturn(new GitRepositoryImpl());
```
shouldn't the url be the one you set in line 46?

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryService.java:
```diff
@@ -0,0 +1,83 @@
+        
+        try{
+            jGitService.clone(path + requester + "_" + owner + "_" + repositoryName, cloneForm.getUrl());
+            gitRepositoryJpa.save(gitRepository);
```
you probably want to return this repository object since it has the id of the new object

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryManagerServiceTest.java:
```diff
@@ -0,0 +1,56 @@
+    @Test
+    public void test_cloneRepository_clone() {
+        Mockito.when(gitRepositoryJpa.findByUrl("")).thenReturn(new GitRepositoryImpl());
+        Assertions.assertDoesNotThrow(()->serviceToTest.cloneRepository(cloneForm, user));
```
when this method returns a repository object, you should check that.

jormsby2 replied:
Can you explain a bit more what you meant here? Should I be checking local variable values somehow?

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryManagerServiceTest.java:
```diff
@@ -0,0 +1,56 @@
+    public void test_cloneRepository_clone() {
+        Mockito.when(gitRepositoryJpa.findByUrl("")).thenReturn(new GitRepositoryImpl());
+        Assertions.assertDoesNotThrow(()->serviceToTest.cloneRepository(cloneForm, user));
+    }   
```
you also need a test case for when `jGitService.clone` throws an exception

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryManagerServiceTest.java:
```diff
@@ -0,0 +1,56 @@
+        Assertions.assertDoesNotThrow(()->serviceToTest.cloneRepository(cloneForm, user));
+    }   
+    
+}
```
the other methods need test cases too (list and delete)

jormsby2 committed:
[STC-5] Refactored services to accommodate any remote repository

jormsby2 commented:
I said refactored in my commit but that might not be the right word to use here, I changed the services and entities pretty significantly such that they can accommodate any remote repository. 

I also added a service to create folder names from a url.

jormsby2 commented:
I'm not really sure what I did wrong but all my commits are showing as oct 26, but the author dates are still correct. I think it might be because I had rebase selected instead of merge when I pushed?

jdamerow commented:
It looks like you force-push your changes, which is usually not a good idea and not really necessary either. When did you rebase? Did you merge develop into your branch?

jormsby2 commented:
> force-push

I was trying to clean up my commits by rebasing my branch against develop, I've done this in the past and never had issues so I think the issue must be that I force pushed? I'll set aside some time to study git more closely as I don't seem to understand as well as I thought I did.

jdamerow commented:
Yes, that was probably the issue. If you rebase properly I think you usually don’t have to force push. GitHub’s recommendation that we typically follow though is to merge instead of rebase ( I can’tfind the source right now but at least at some point it was what they suggested). Especially once you push your branch rebasing can make a mess if others have checked out your branch. 

Also, I usually squash the commits when I merge a PR anyways.

Sent from my iPhone

> On Oct 27, 2020, at 15:58, Jason Ormsby <notifications@github.com> wrote:
> 
> ﻿
> force-push
> 
> I was trying to clean up my commits by rebasing my branch against develop, I've done this in the past and never had issues so I think the issue must be that I force pushed? I'll set aside some time to study git more closely as I don't seem to understand as well as I thought I did.
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub, or unsubscribe.

jormsby2 committed:
[STC-15] Updated jgit service to make tar of cloned repository

jormsby2 committed:
Revert "[STC-15] Updated jgit service to make tar of cloned repository"

This reverts commit d37b32e37fd21dd15cdb245de2f91e61a584ec0b.

jormsby2 committed:
Revert "[STC-5] Refactored services to accommodate any remote repository"

This reverts commit ed8181c665d07d3d750a814be3589fef5a211f03.

jormsby2 committed:
Revert "Revert "[STC-5] Refactored services to accommodate any remote repository""

This reverts commit 42b5f7b09d1f4472076f92e3165525b86e1fd886.

jormsby2 commented:
I have a service that deletes files and one that makes folder names from url's, I just realized after looking at this branch again I should probably merge these together into a single file centric service?

jormsby2 committed:
[STC-5] Formatting

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DeleteFilesServiceImplTest.java:
```diff
@@ -0,0 +1,47 @@
+    
+    @Test
+    public void test_deleteDirectoryContents_noSuchFile() throws IOException{
+    //file file = null
```
remove

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryManagerServiceTest.java:
```diff
@@ -0,0 +1,73 @@
+    @Test
+    public void test_cloneRepository_clone() throws MalformedURLException {
+        Mockito.when(urlFormatter.urlToFolderName("https://github.com/diging/scriptoocloud"))
+            .thenReturn("github.com_diging_scriptoocloud");
```
I would have  the formatter service replace `.` with `_`. Some os might interpret it as an extension otherwise and make it harder to work with (not a huge deal, but might as well, since you have control over that).

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryManagerServiceTest.java:
```diff
@@ -0,0 +1,73 @@
+            .thenReturn("github.com_diging_scriptoocloud");
+        Mockito.when(gitRepositoryJpa.findByUrl("github.com_diging_scriptoocloud")).thenReturn(null);
+        Assertions.assertDoesNotThrow(
+            ()->serviceToTest.cloneRepository("https://github.com/diging/scriptoocloud", "admin"));     
```
I would verify that the jitservice is called and that the save method is called.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryManagerServiceTest.java:
```diff
@@ -0,0 +1,73 @@
+    }
+        
+    @Test
+    public void test_cloneRepository_list() {
```
what does this method test? The `cloneRepository` method? No, right? The test method name should be `test_` + method to be tested + `_` + scenario tested.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryManagerServiceTest.java:
```diff
@@ -0,0 +1,73 @@
+    public void test_cloneRepository_list() {
+        Mockito.when(gitRepositoryJpa.findAll()).thenReturn((Iterable)new ArrayList<GitRepository>());
+        Assertions.assertEquals(new ArrayList<GitRepository>(), serviceToTest.listRepositories());
+    }   
```
needs a test where there are repository objects being returned.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryManagerServiceTest.java:
```diff
@@ -0,0 +1,73 @@
+        gitRepositoryImpl.setFolderName("test");
+        Mockito.when(gitRepositoryJpa.findById((long) 0))
+                .thenReturn(Optional.of(gitRepositoryImpl));
+        Assertions.assertDoesNotThrow(()->serviceToTest.deleteRepository((long) 0));
```
You should verify that the delete method that deletes the contents is being called and that the delete method on the repo is called.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/JgitServiceImplTest.java:
```diff
@@ -0,0 +1,43 @@
+        Assertions.assertThrows(InvalidGitUrlException.class,
+            ()->serviceToTest.clone("testFile", "githubm/test"));
+    }   
+}
```
needs a test case for when everything works fine.

jdamerow commented:
> I have a service that deletes files and one that makes folder names from url's, I just realized after looking at this branch again I should probably merge these together into a single file centric service?

I think it's fine to have to classes; I would just rename the formatter class as `Utility` rather than `Service`, since it really just does some utility work.

jormsby2 committed:
[STC-5] pull request chagnes

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneGitRepositoriesController.java:
```diff
@@ -0,0 +1,71 @@
+    @Autowired
+    private GitRepositoryManager gitRepositoryManager;
+    
+    @Value("${git.repositories.path}")
```
Is this working? I'd expect a `@ProperySource` annotation on the class for this to work?

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneGitRepositoriesController.java:
```diff
@@ -0,0 +1,71 @@
+        if(result.hasErrors()){
+            model.addAttribute("clone",cloneForm);
+            redirectAttributes.addAttribute("formResponse","No Url provided");
+            return "redirect:/repositories/clone";
```
For validation errors, the errors should already be in the model, you just need to return the page  that shows the form (`repositories/clone`, no redirecting) and potentially put the cloneForm into the model so that the previously entered values are available.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneGitRepositoriesController.java:
```diff
@@ -0,0 +1,71 @@
+            logger.error("Invalid Url provided " + cloneForm.getUrl());
+            redirectAttributes.addAttribute("formResponse","Invalid Url");
+            model.addAttribute("clone",cloneForm);            
+            return "redirect:/repositories/clone";
```
I would use a similar approach to above here, return the form page, put the cloneForm into the model and either add an error message to the binding results or put an error message in the model. Malformed URLs could (and probably should) be validated with a validator to avoid this issue here.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/JgitServiceImplTest.java:
```diff
@@ -0,0 +1,50 @@
+    
+    @Test
+    public void test_clone_success(){
+        Assertions.assertDoesNotThrow(()->serviceToTest.clone("testCloneSuccess", "https://github.com/diging/scriptoocloud"));
```
I wouldn't use a relative folder here, but make that configurable. You should be able to create a properties object from a file in `src/test/resources` and then use the configured location here (similar to how it is done in the actual code). You also want to make sure your that you clean up after your tests (removing the created directory, so it doesn't interfere with your next test).

jdamerow replied:
Last but not least, does it make a difference if the folder where the repository should be cloned has contents? Since Git doesn't like it if you do it on the command line, I assume it does? In that case, you should have a test case for both scenarios.

jormsby2 replied:
Can I remove this clone success test since it's just the standard JGit clone call? I've spent a lot of time trying to figure out how to kill the JGit clone process and release file system locks its holding and I'm not having any success. This results in my cleanup call in @AfterEach only partially cleaning up the folder created. If this test is necessary the only solution I can think of is modifying my JGit service to accommodate tests better?

jdamerow replied:
Let's talk about it in our meeting today, I think I don't understand the problems you are running into.

jormsby2 committed:
[STC-5] Unit tests now delete directories

jormsby2 committed:
[STC-5] custom validator

jormsby2 committed:
[STC-5] Custom validation controller and view

jormsby2 committed:
[STC-5] validation fix

jormsby2 committed:
[STC-5] Formatting

jormsby2 committed:
[STC-5] test paths configurable from src/test/config.properties

jormsby2 committed:
[STC-5] Jgit test fix

jormsby2 committed:
[STC-5] Finished unit tests

jormsby2 committed:
Merge branch 'develop' into task/STC-5

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/config/SecurityContext.java:
```diff
@@ -37,7 +37,7 @@ protected void configure(HttpSecurity http) throws Exception {
                         "/register").permitAll()
                 // The rest of the our application is protected.
-                .antMatchers("/users/**", "/admin/**").hasRole("ADMIN")
+                .antMatchers("/repositories/**", "/users/**", "/admin/**").hasRole("ADMIN")
```
if urls under `/repositories` are only usable by admin I would put them with the prefix `/admin/` and then you don't have to change anything here.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryService.java:
```diff
@@ -0,0 +1,81 @@
+        GitRepository repositoryEntity = gitRepositoryJpa.findByUrl(gitUrl);
+        
+        if(repositoryEntity != null){
+            deleteRepository(repositoryEntity.getId());
```
hm, I wouldn't delete the repository object here, but use the existing one and update it. You'll probably have dependencies to the repository objects later (e.g. from the projects that use a repo) and then you either wouldn't be able to delete the repo here without removing the relationship from projects to repo, or you would have links from projects to repo objects that don't exist anymore.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/JgitServiceImpl.java:
```diff
@@ -0,0 +1,37 @@
+    private DeleteFilesService deleteFilesService;
+
+    @Override
+    public void clone(String filename, String url) throws InvalidGitUrlException{
```
what is `filename`? the name of the directory to clone to? needs a better variable name. Also, this class needs javadoc.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/UrlFormatterUtilityImpl.java:
```diff
@@ -0,0 +1,35 @@
+        String folderName = "";
+        
+        Scanner scanner = new Scanner(urlWrapper.getHost()).useDelimiter("\\.");
+        folderName += scanner.next();
```
merge with line 18

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneGitRepositoriesController.java:
```diff
@@ -0,0 +1,63 @@
+    
+    @Autowired
+    private GitRepositoryManager gitRepositoryManager;
+    private Logger logger = LoggerFactory.getLogger(getClass());    
```
this should be the first declaration in the class and final.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DeleteFilesServiceImplTest.java:
```diff
@@ -0,0 +1,54 @@
+        File file = folder.resolve("myfile").toFile();
+        Assertions.assertEquals(true, file.exists());
+        serviceToTest.deleteDirectoryContents(file);
+        Assertions.assertEquals(false, file.exists());
```
this test doesn't seem to test folder deletion, does it? Just a file with some content?

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryManagerServiceTest.java:
```diff
@@ -0,0 +1,106 @@
+            ()->serviceToTest.cloneRepository("https://github.com/diging/scriptoocloud", "admin"));  
+            
+        GitRepositoryImpl repo  = new GitRepositoryImpl();
+        Mockito.when(gitRepositoryJpa.save(repo)).thenReturn(repo);
```
where is that needed that you define it after calling the method to test?

jormsby2 committed:
[STC-5] Update DB entity instead of removing and adding

jormsby2 committed:
[STC-5] Moved repository routes to admin folder

jormsby2 committed:
[STC-5] Added javadoc to services

jormsby2 committed:
[STC-5] Unit test was missing assertion for deleted folder

jormsby2 committed:
[STC-5] Formatting

jormsby2 committed:
[STC-5] Formatting

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryService.java:
```diff
@@ -0,0 +1,110 @@
+            repositoryEntity.setUrl(gitUrl);
+            repositoryEntity.setRequester(requester);
+            repositoryEntity.setCreationDate(creationDate);
+            repositoryEntity.setFolderName(folderName);
```
there are several lines of duplicate code here, should be DRY (don't repeat yourself)

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/JgitServiceImpl.java:
```diff
@@ -0,0 +1,50 @@
+            throw new InvalidGitUrlException(e);
+        } catch(JGitInternalException e){
+            deleteFilesService.deleteDirectoryContents(new File(localRepoFolderName));
+            throw new InvalidGitUrlException(e);
```
hm, I would rename this exception or use a different one. The exceptions thrown by jgit can be for many reasons not just invalid urls, can they? Also, if you handle both exceptions the same way, you can catch them in the same block `GitAPIException|JGitInternalException e`

jdamerow replied:
E.g. I forgot to set up the repositories path, and when I tried to clone a repo I got an invalid url error message which didn't make sense.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneGitRepositoriesController.java:
```diff
@@ -0,0 +1,65 @@
+        } catch(InvalidGitUrlException e){
+            logger.error("No git repository found at provided URL " + cloneForm.getUrl());
+            redirectAttributes.addAttribute("formResponse","No such git repository found");
+            model.addAttribute("clone",cloneForm);
```
are you sure this works? since you're redirecting, I think you would loose your model attributes.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/CloneGitRepositoriesController.java:
```diff
@@ -0,0 +1,65 @@
+            return "redirect:/admin/repositories/clone";
+        } catch(MalformedURLException e){
+            logger.error("Malformed URL made it past validator " + cloneForm.getUrl());
+            return "/admin/repositories/clone";
```
Don't you want to the use to correct their entries in this case (so previous values should be shown)?

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/validation/RepositoryUrlConstraintValidator.java:
```diff
@@ -0,0 +1,28 @@
+            URL url = new URL(urlField);
+        } catch(final Exception ignore){
+            logger.error("Malformed URL provided");
+            valid = false;    
```
return false

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/validation/RepositoryUrlConstraintValidator.java:
```diff
@@ -0,0 +1,28 @@
+            logger.error("Malformed URL provided");
+            valid = false;    
+        }
+        return valid;
```
return true

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/validation/RepositoryUrlConstraintValidator.java:
```diff
@@ -0,0 +1,28 @@
+    public void initialize(RepositoryUrlConstraint constraintAnnotation) {}
+
+    public boolean isValid(String urlField, ConstraintValidatorContext context) {
+        boolean valid = true;
```
remove

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/validation/RepositoryUrlConstraintValidator.java:
```diff
@@ -0,0 +1,28 @@
+        boolean valid = true;
+        try {
+            URL url = new URL(urlField);
+        } catch(final Exception ignore){
```
why is that final?

jdamerow replied:
why do you catch `Exception` here? It's throwing a `MalformedURLException`, doesn't it?

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/validation/RepositoryUrlConstraintValidator.java:
```diff
@@ -0,0 +1,28 @@
+        try {
+            URL url = new URL(urlField);
+        } catch(final Exception ignore){
+            logger.error("Malformed URL provided");
```
log stack trace and make it warn or info. It's a validator, so it's job is to find non-valid inputs. It doesn't need to log in error level, but when you log it you want the stacktrace in case some comes and asks "what's wrong with this".

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/JgitServiceImplTest.java:
```diff
@@ -0,0 +1,99 @@
+        Assertions.assertTrue(file.exists());
+        
+        DeleteFilesServiceImpl deleteFilesService = new DeleteFilesServiceImpl();
+        deleteFilesService.deleteDirectoryContents(file);
```
hm, I understand why you do this for convenience, but I would just add a method to this class to delete a folder. If someone changes DeleteFilesServiceImpl and introduces bugs, not just the tests for DeleteFilesServiceImpl would fail, but also these test cases here, which would make it more confusing to figure out what is wrong.

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
Make it so, Jenkins.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryService.java:
```diff
@@ -0,0 +1,110 @@
+            repositoryEntity.setRequester(requester);
+            repositoryEntity.setCreationDate(creationDate);
+            repositoryEntity.setFolderName(folderName);
+            File file = new File(path + repositoryEntity.getFolderName());
```
you probably want to make sure path ends with `/` or this will fail

jormsby2 replied:
I've been looking for a good solution for this and I see that it is possible to use groovy plugin to run scripts in pom file. Does this seem like a good solution or do you have a better way I could deal with validity checks on pom property before it's injected? I'm assuming checking validity after injection would be a bad idea since many classes would need to run independent validity checks to ensure pattern of ending with "/"?

jdamerow replied:
I think you're overthinking it. How often is the `path` property used? I would either do the check after object construction `@PostConstruct` or if it is used a lot, I would make it available as a bean and do the check in the method that creates the bean `@Bean` annotation in a config class.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!

jormsby2 committed:
[STC-5] conform to DRY, check for trailing slash, validator fix

jormsby2 committed:
[STC-5] some flash attributes changed, fixed tests, fixed excpeitons

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/DeleteFilesServiceImpl.java:
```diff
@@ -0,0 +1,37 @@
+            }
+            file.delete();
+        }
+        else{
```
please follow our style guide (you might want to change that in your eclipse settings)

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryService.java:
```diff
@@ -0,0 +1,118 @@
+    */
+    @PostConstruct
+    private void validatePathProperty(){
+        if(!path.substring(path.length()-1).equals("/")){
```
you should probably use `endsWith()` to make it more readable

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/GitRepositoryService.java:
```diff
@@ -0,0 +1,118 @@
+        if(repositoryEntity == null){
+            repositoryEntity = new GitRepositoryImpl();
+        }
+        else{
```
code style

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Build failed. Maybe you broke the test cases?

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Build failed. Maybe you broke the test cases?

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!

jormsby2 committed:
[STC-5] Styling,readability

jormsby2 committed:
[STC-5] Styling

jormsby2 commented:
Thank you for the endsWith() suggestion! I can't believe I've never seen that method before, it looks much nicer. I think the styling should be good now? I tried comparing all my files to the styling in citesphere to see what I did wrong in that regard.

jdamerow commented:
Make it so, Jenkins.

jdamerow commented:
This is the style guide we're following (except the indentation setting):
https://google.github.io/styleguide/javaguide.html

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
For the UI: I would put the clone button either on the manage repository site at the top somewhere or make "manage repositories" a drop down menu with two entries.

jormsby2 committed:
[STC-5] Integrated FileSystemService, added dropdown

jormsby2 commented:
Converted my menu items to drop down menu. Removed my file service and now use john's FileSystemService for deletions.

diging-jenkins commented:
Congratulations, build succeeded!
